### ISSUE TITLE
Parsing of section 3par_node failed

### ISSUE THREAD

**Issue by craftsolution** [Yogibaer75/Check_MK-Things#75]
Hello,

I have tried to use the 3par_ssh package and the check is not collecting data, it endup with this warning:

```
WARNING: Parsing of section 3par_node failed - please submit a crash report! (Crash-ID: 8792c78a-773a-11ef-be2f-ded04cdc0b7e)
Job details
Starting job...
WARNING: Parsing of section 3par_node failed - please submit a crash report! (Crash-ID: 8792c78a-773a-11ef-be2f-ded04cdc0b7e)
Completed.
```

my system: HPE Alletra 9000

Do you know what is wrong ? 

Best regards,

**Comment by Yogibaer75** [comment]
In this case i need the raw agent output from your device. As the failure happens at parsing time, it means the data can be collected from the device.

**Comment by craftsolution** [comment]
> In this case i need the raw agent output from your device. As the failure happens at parsing time, it means the data can be collected from the device.

Hello,
Yes, the data are collected but the parsing is not done correctly.

I had this as the output
```
<<<3par_node>>>
   0 CZ23080CG7-0 OK      No     Yes       Off          GreenBlnk  210624   51200          100
   1 CZ23080CG7-1 OK      Yes    Yes       Off          GreenBlnk  210624   51200          100
```
I have  changed the parsing code by using a simple mapping by attributing each variables extract from ```ThreeParNode``` by using node_id as a key to store each entry from ThreeParNode.
```
def parse_3par_node(string_table: StringTable) -> Section:
    """parse raw data into dict"""
    parsed: dict[str, ThreeParNode] = {}
    for line in string_table:
        # Assurer que la ligne a bien 10 colonnes
        if len(line) == 10 and line[0].isdigit():
            (
                node_id,
                name,
                state,
                master,
                cluster,
                service_led,
                master_led,
                control_mem,
                data_mem,
                cache,
            ) = line
            parsed[node_id] = ThreeParNode(
                node_id=node_id,
                name=name,
                state=state,
                master=master,
                cluster=cluster,
                service_led=service_led,
                master_led=master_led,
                control_mem=control_mem,
                data_mem=data_mem,
                cache=cache,
            )
    return parsed
```

There is also a problem with 3par_pd.py script. I have not correct it yet.
To gon on for testing the ruleset "[HPE 3par with SSH]" I  have disabled 3par_pd in special_agents/agent_threepar_ssh.py but when validating the new monitored services I have an error message du to the fact that your are using number in scripts names.
![image](https://github.com/user-attachments/assets/cd5e6619-cb0b-4f7f-acf5-65b8c0a1f817)

```
Started at: 14:48:28. Finished at: 14:48:30.
Error running automation call restart (exit code 5), error:
Error precompiling checks for host par01:   File "/omd/sites/cmk/var/check_mk/core/helper_config/112/host_checks/para01", line 40
    import cmk_addons.plugins.threepar_ssh.agent_based.3par_node
                                                       ^
SyntaxError: invalid decimal literal
```
This error is du to the fact you are using number in script names, you might change it to 3par_node.py -> threepar_node.py
![image](https://github.com/user-attachments/assets/2d880daf-7317-42d0-a388-7be1537d489f)

If you need the output for 3par_pd, here it is
```
<<<3par_pd>>>
                           ----Size(MiB)---- -----Ports-----
Id CagePos Type RPM State     Total     Free A       B       Capacity(GB)
 0 0:0:0   SSD  N/A normal  3660800  1294336 0:11:1* 1:11:1*         3840
 1 0:1:0   SSD  N/A normal  3660800  1319936 0:12:1* 1:12:1*         3840
 2 0:2:0   SSD  N/A normal  3660800  1295360 0:13:1* 1:13:1*         3840
 3 0:3:0   SSD  N/A normal  3660800  1318912 0:14:1* 1:14:1*         3840
 4 0:4:0   SSD  N/A normal  3660800  1295360 0:15:1* 1:15:1*         3840
 5 0:5:0   SSD  N/A normal  3660800  1321984 0:16:1* 1:16:1*         3840
 6 0:6:0   SSD  N/A normal  3660800  1293312 0:17:1* 1:17:1*         3840
 7 0:7:0   SSD  N/A normal  3660800  1319936 0:18:1* 1:18:1*         3840
 8 0:8:0   SSD  N/A normal  3660800  1294336 0:19:1* 1:19:1*         3840
 9 0:9:0   SSD  N/A normal  3660800  1320960 0:20:1* 1:20:1*         3840
10 0:10:0  SSD  N/A normal  3660800  1293312 0:21:1* 1:21:1*         3840
11 0:11:0  SSD  N/A normal  3660800  1321984 0:22:1* 1:22:1*         3840
12 2:0:0   SSD  N/A normal  3660800  1294336 0:0:1*  1:0:1*          3840
13 2:1:0   SSD  N/A normal  3660800  1320960 0:0:1*  1:0:1*          3840
14 2:2:0   SSD  N/A normal  3660800  1295360 0:0:1*  1:0:1*          3840
15 2:3:0   SSD  N/A normal  3660800  1321984 0:0:1*  1:0:1*          3840
16 2:4:0   SSD  N/A normal  3660800  1295360 0:0:1*  1:0:1*          3840
17 2:5:0   SSD  N/A normal  3660800  1321984 0:0:1*  1:0:1*          3840
18 2:6:0   SSD  N/A normal  3660800  1293312 0:0:1*  1:0:1*          3840
19 2:7:0   SSD  N/A normal  3660800  1319936 0:0:1*  1:0:1*          3840
20 2:8:0   SSD  N/A normal  3660800  1295360 0:0:1*  1:0:1*          3840
21 2:9:0   SSD  N/A normal  3660800  1321984 0:0:1*  1:0:1*          3840
22 2:10:0  SSD  N/A normal  3660800  1294336 0:0:1*  1:0:1*          3840
23 2:11:0  SSD  N/A normal  3660800  1321984 0:0:1*  1:0:1*          3840
-------------------------------------------------------------------------
24 total                   87859200 31386624
```


Best regards,

**Comment by OlaHagbohm** [comment]
Apologies to ressurect an old issue here, but I was facing the same issues as olexdev113.
I updated 3par_vv.py, 3par_node.py and so on according to olexdev113´s comment.

3par_vv.py:
```
from typing import Mapping
from attr import dataclass
from cmk.agent_based.v2 import (
    AgentSection,
    CheckPlugin,
    CheckResult,
    DiscoveryResult,
    Result,
    Service,
    State,
    StringTable,
)


@dataclass(frozen=True)
class ThreeParVV:
    """one vv"""
    vv_id: str
    name: str
    prov: str
    vv_type: str
    copyof: str
    base_id: str
    read_write: str
    state: str
    rsvd_adm: str
    rsvd_snp: str
    rsvd_usr: str
    vsize: str


Section = Mapping[str, ThreeParVV]


def parse_3par_vv(string_table: StringTable) -> Section:
    """parse raw data into dict"""
    parsed: dict[str, ThreeParVV] = {}
    for line in string_table:
        if len(line) == 12 and line[0].isdigit():
            (vv_id,name,prov,vv_type,copyof,base_id,read_write,state,rsvd_adm,rsvd_snp,rsvd_usr,vsize) = line
            parsed[vv_id] = ThreeParVV(vv_id,name,prov,vv_type,copyof,base_id,read_write,state,rsvd_adm,rsvd_snp,rsvd_usr,vsize)
    return parsed


agent_section_3par_vv = AgentSection(
    name="3par_vv",
    parse_function=parse_3par_vv,
    parsed_section_name="3par_vv",
)


def discovery_3par_vv(section: Section) -> DiscoveryResult:
    """if data is present return one service per vv"""
    if section:
        for element in section:
            yield Service(item=element)


def check_3par_vv(item: str, section: Section) -> CheckResult:
    """check the vv state"""
    data = section.get(item)
    if not data:
        return

    # _reserved_space = int(data.rsvd_adm) + int(data.rsvd_snp) + int(data.rsvd_usr)
    message = f"VV {data.vv_id.zfill(3)}/{data.name} ({data.vsize} MB) has status {data.state}"
    if data.state == "normal":
        yield Result(state=State.OK, summary=message)
    else:
        yield Result(state=State.CRIT, summary=message + "(!!)")


check_plugin_3par_vv = CheckPlugin(
    name="3par_vv",
    service_name="VV %s",
    sections=["3par_vv"],
    discovery_function=discovery_3par_vv,
    check_function=check_3par_vv,
)
```


Discovery in the GUI crashes with a value error

> ValueError (not enough values to unpack (expected 12, got 1))

Traceback:
```
  File "/omd/sites/monitor/lib/python3/cmk/checkengine/sectionparser.py", line 121, in _parse_raw_data
    return parse_function(list(raw_data))
  File "/omd/sites/monitor/local/lib/python3/cmk_addons/plugins/threepar_ssh/agent_based/3par_vv.py", line 57, in parse_3par_vv
    (vv_id,name,prov,vv_type,copyof,base_id,read_write,state,rsvd_adm,rsvd_snp,rsvd_usr,vsize) = line
not enough values to unpack (expected 12, got 1)
```

Local Variables:
```
{'line': ['2',
          '.mgmtdata',
          'full',
          'NA',
          'NA',
          'base',
          '---',
          '2',
          'RW',
          'normal',
          '524288',
          '524288'],
 'parsed': {},
 'string_table': [['Id',
                   'Name',
                   'Prov',
                   'Compr',
                   'Dedup',
                   'Type',
                   'CopyOf',
                   'BsId',
                   'Rd',
                   '-Detailed_State-',
                   'Rsvd(MiB)',
                   'VSize(MiB)'],
                  ['2',
                   '.mgmtdata',
                   'full',
                   'NA',
                   'NA',
                   'base',
                   '---',
                   '2',
                   'RW',
                   'normal',
                   '524288',
                   '524288'],
                  ['26',
                   '.shared.SSD_r6_0',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '26',
                   'RW',
                   'normal',
                   '60375',
                   '4194304'],
                  ['27',
                   '.shared.SSD_r6_1',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '27',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['28',
                   '.shared.SSD_r6_2',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '28',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['29',
                   '.shared.SSD_r6_3',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '29',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['30',
                   '.shared.SSD_r6_4',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '30',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['31',
                   '.shared.SSD_r6_5',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '31',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['32',
                   '.shared.SSD_r6_6',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '32',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['33',
                   '.shared.SSD_r6_7',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '33',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['34',
                   '.shared.SSD_r6_8',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '34',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['35',
                   '.shared.SSD_r6_9',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '35',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['36',
                   '.shared.SSD_r6_a',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '36',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['37',
                   '.shared.SSD_r6_b',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '37',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['38',
                   '.shared.SSD_r6_c',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '38',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['39',
                   '.shared.SSD_r6_d',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '39',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['40',
                   '.shared.SSD_r6_e',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '40',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['41',
                   '.shared.SSD_r6_f',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '41',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['44',
                   'H2H1-Alletra-V002',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '44',
                   'RW',
                   'normal',
                   '452025',
                   '5242880'],
                  ['45',
                   'H1H2-Alletra-V001',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '45',
                   'RW',
                   'normal',
                   '99225',
                   '5242880'],
                  ['46',
                   'H1H2-Alletra-V003',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '46',
                   'RW',
                   'normal',
                   '640500',
                   '5242880'],
                  ['47',
                   'H2H1-Alletra-V004',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '47',
                   'RW',
                   'normal',
                   '215250',
                   '5242880'],
                  ['48',
                   'H1-Alletra-V100',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '48',
                   'RW',
                   'normal',
                   '6825',
                   '1048576'],
                  ['67',
                   'H1H2-Alletra-V005',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '67',
                   'RW',
                   'normal',
                   '772800',
                   '5242880'],
                  ['68',
                   'H1H2-Alletra-V007',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '68',
                   'RW',
                   'normal',
                   '326025',
                   '5242880'],
                  ['69',
                   'H1H2-Alletra-V009',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '69',
                   'RW',
                   'normal',
                   '266175',
                   '5242880'],
                  ['70',
                   'H1H2-Alletra-V011',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '70',
                   'RW',
                   'normal',
                   '280875',
                   '5242880'],
                  ['71',
                   'H1H2-Alletra-V013',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '71',
                   'RW',
                   'normal',
                   '1833300',
                   '5242880'],
                  ['72',
                   'H2H1-Alletra-V006',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '72',
                   'RW',
                   'normal',
                   '287175',
                   '5242880'],
                  ['73',
                   'H2H1-Alletra-V008',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '73',
                   'RW',
                   'normal',
                   '580125',
                   '5242880'],
                  ['74',
                   'H2H1-Alletra-V010',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '74',
                   'RW',
                   'normal',
                   '405825',
                   '5242880'],
                  ['75',
                   'H2H1-Alletra-V012',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '75',
                   'RW',
                   'normal',
                   '1249500',
                   '5242880'],
                  ['76',
                   'H2H1-Alletra-V014',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '76',
                   'RW',
                   'normal',
                   '181125',
                   '5242880'],
                  ['79',
                   'H1H2-scratchISO-V000',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '79',
                   'RW',
                   'normal',
                   '1050',
                   '1048576'],
                  ['1',
                   '.srdata',
                   'full',
                   'NA',
                   'NA',
                   'base',
                   '---',
                   '1',
                   'RW',
                   'normal',
                   '92160',
                   '92160'],
                  ['0',
                   'admin',
                   'full',
                   'NA',
                   'NA',
                   'base',
                   '---',
                   '0',
                   'RW',
                   'normal',
                   '10240',
                   '10240'],
                  ['---------------------------------------------------------------------------------------------------'],
                  ['35', 'total', '9178413', '143233024'],
                  ['Id',
                   'Name',
                   'Prov',
                   'Compr',
                   'Dedup',
                   'Type',
                   'CopyOf',
                   'BsId',
                   'Rd',
                   '-Detailed_State-',
                   'Rsvd(MiB)',
                   'VSize(MiB)'],
                  ['2',
                   '.mgmtdata',
                   'full',
                   'NA',
                   'NA',
                   'base',
                   '---',
                   '2',
                   'RW',
                   'normal',
                   '524288',
                   '524288'],
                  ['26',
                   '.shared.SSD_r6_0',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '26',
                   'RW',
                   'normal',
                   '60375',
                   '4194304'],
                  ['27',
                   '.shared.SSD_r6_1',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '27',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['28',
                   '.shared.SSD_r6_2',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '28',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['29',
                   '.shared.SSD_r6_3',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '29',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['30',
                   '.shared.SSD_r6_4',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '30',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['31',
                   '.shared.SSD_r6_5',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '31',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['32',
                   '.shared.SSD_r6_6',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '32',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['33',
                   '.shared.SSD_r6_7',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '33',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['34',
                   '.shared.SSD_r6_8',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '34',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['35',
                   '.shared.SSD_r6_9',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '35',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['36',
                   '.shared.SSD_r6_a',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '36',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['37',
                   '.shared.SSD_r6_b',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '37',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['38',
                   '.shared.SSD_r6_c',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '38',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['39',
                   '.shared.SSD_r6_d',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '39',
                   'RW',
                   'normal',
                   '59325',
                   '4194304'],
                  ['40',
                   '.shared.SSD_r6_e',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '40',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['41',
                   '.shared.SSD_r6_f',
                   'dds',
                   'v2',
                   'NA',
                   'base',
                   '---',
                   '41',
                   'RW',
                   'normal',
                   '59850',
                   '4194304'],
                  ['44',
                   'H2H1-Alletra-V002',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '44',
                   'RW',
                   'normal',
                   '452025',
                   '5242880'],
                  ['45',
                   'H1H2-Alletra-V001',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '45',
                   'RW',
                   'normal',
                   '99225',
                   '5242880'],
                  ['46',
                   'H1H2-Alletra-V003',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '46',
                   'RW',
                   'normal',
                   '640500',
                   '5242880'],
                  ['47',
                   'H2H1-Alletra-V004',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '47',
                   'RW',
                   'normal',
                   '215250',
                   '5242880'],
                  ['48',
                   'H1-Alletra-V100',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '48',
                   'RW',
                   'normal',
                   '6825',
                   '1048576'],
                  ['67',
                   'H1H2-Alletra-V005',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '67',
                   'RW',
                   'normal',
                   '772800',
                   '5242880'],
                  ['68',
                   'H1H2-Alletra-V007',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '68',
                   'RW',
                   'normal',
                   '326025',
                   '5242880'],
                  ['69',
                   'H1H2-Alletra-V009',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '69',
                   'RW',
                   'normal',
                   '266175',
                   '5242880'],
                  ['70',
                   'H1H2-Alletra-V011',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '70',
                   'RW',
                   'normal',
                   '280875',
                   '5242880'],
                  ['71',
                   'H1H2-Alletra-V013',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '71',
                   'RW',
                   'normal',
                   '1833300',
                   '5242880'],
                  ['72',
                   'H2H1-Alletra-V006',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '72',
                   'RW',
                   'normal',
                   '287175',
                   '5242880'],
                  ['73',
                   'H2H1-Alletra-V008',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '73',
                   'RW',
                   'normal',
                   '580125',
                   '5242880'],
                  ['74',
                   'H2H1-Alletra-V010',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '74',
                   'RW',
                   'normal',
                   '405825',
                   '5242880'],
                  ['75',
                   'H2H1-Alletra-V012',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '75',
                   'RW',
                   'normal',
                   '1249500',
                   '5242880'],
                  ['76',
                   'H2H1-Alletra-V014',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '76',
                   'RW',
                   'normal',
                   '181125',
                   '5242880'],
                  ['79',
                   'H1H2-scratchISO-V000',
                   'tdvv',
                   'v2',
                   'Yes',
                   'base',
                   '---',
                   '79',
                   'RW',
                   'normal',
                   '1050',
                   '1048576'],
                  ['1',
                   '.srdata',
                   'full',
                   'NA',
                   'NA',
                   'base',
                   '---',
                   '1',
                   'RW',
                   'normal',
                   '92160',
                   '92160'],
                  ['0',
                   'admin',
                   'full',
                   'NA',
                   'NA',
                   'base',
                   '---',
                   '0',
                   'RW',
                   'normal',
                   '10240',
                   '10240'],
                  ['---------------------------------------------------------------------------------------------------'],
                  ['35', 'total', '9178413', '143233024']]}
```



Discovery using cmk --discover -vvv command output: 

```
Discovering services and host labels on: servername.domain.local
hostname.infra.network:
+ FETCHING DATA
  Source: SourceInfo(hostname='servername.domain.local', ipaddress='10.1.2.80', ident='special_threepar_ssh', fetcher_type=<FetcherType.SPECIAL_AGENT: 6>, source_type=<SourceType.HOST: 1>)
[cpu_tracking] Start [7096b601d760]
Read from cache: AgentFileCache(path_template=/omd/sites/monitor/tmp/check_mk/data_source_cache/special_threepar_ssh/servername.domain.local, max_age=MaxAge(checking=0, discovery=90.0, inventory=90.0), simulation=False, use_only_cache=False, file_cache_mode=1)
Calling: /omd/sites/monitor/local/lib/python3/cmk_addons/plugins/threepar_ssh/libexec/agent_threepar_ssh -u monitor -i showcage,showpd,showld,showvv,showps,shownode 10.1.2.80
Get data from program
[cpu_tracking] Stop [7096b601d760 - Snapshot(process=posix.times_result(user=0.0, system=0.0, children_user=0.13, children_system=0.02, elapsed=3.8200000002980232))]
+ PARSE FETCHER RESULTS
<<<3par_cage>>> / Transition NOOPParser -> HostSectionParser
<<<3par_pd>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_ld>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_vv>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_ps>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_node>>> / Transition HostSectionParser -> HostSectionParser
  HostKey(hostname='servername.domain.local', source_type=<SourceType.HOST: 1>)  -> Add sections: ['3par_cage', '3par_ld', '3par_node', '3par_pd', '3par_ps', '3par_vv']
Received no piggyback data
+ ANALYSE DISCOVERED HOST LABELS
Trying host label discovery with: 3par_cage, 3par_ld, 3par_node, 3par_pd, 3par_ps, 3par_vv
Trying host label discovery with:
SUCCESS - Found no new host labels
+ ANALYSE DISCOVERED SERVICES
+ EXECUTING DISCOVERY PLUGINS (6)
  Trying discovery with: 3par_cage, 3par_ps, 3par_ld, 3par_node, 3par_vv, 3par_pd
 35 3par_vv
SUCCESS - Found 35 services
```



3par_vv section of agent
```
<<<3par_vv>>>
Id Name                  Prov Compr Dedup Type CopyOf BsId Rd -Detailed_State- Rsvd(MiB) VSize(MiB)
 2 .mgmtdata             full NA    NA    base ---       2 RW normal              524288     524288
26 .shared.SSD_r6_0      dds  v2    NA    base ---      26 RW normal               60900    4194304
27 .shared.SSD_r6_1      dds  v2    NA    base ---      27 RW normal               59325    4194304
28 .shared.SSD_r6_2      dds  v2    NA    base ---      28 RW normal               60375    4194304
29 .shared.SSD_r6_3      dds  v2    NA    base ---      29 RW normal               59325    4194304
30 .shared.SSD_r6_4      dds  v2    NA    base ---      30 RW normal               59850    4194304
31 .shared.SSD_r6_5      dds  v2    NA    base ---      31 RW normal               59850    4194304
32 .shared.SSD_r6_6      dds  v2    NA    base ---      32 RW normal               59850    4194304
33 .shared.SSD_r6_7      dds  v2    NA    base ---      33 RW normal               59850    4194304
34 .shared.SSD_r6_8      dds  v2    NA    base ---      34 RW normal               59850    4194304
35 .shared.SSD_r6_9      dds  v2    NA    base ---      35 RW normal               60375    4194304
36 .shared.SSD_r6_a      dds  v2    NA    base ---      36 RW normal               59850    4194304
37 .shared.SSD_r6_b      dds  v2    NA    base ---      37 RW normal               60375    4194304
38 .shared.SSD_r6_c      dds  v2    NA    base ---      38 RW normal               59850    4194304
39 .shared.SSD_r6_d      dds  v2    NA    base ---      39 RW normal               59850    4194304
40 .shared.SSD_r6_e      dds  v2    NA    base ---      40 RW normal               59850    4194304
41 .shared.SSD_r6_f      dds  v2    NA    base ---      41 RW normal               60375    4194304
44  H2H1-Alletra-V002    tdvv v2    Yes   base ---      44 RW normal              452025    5242880
45  H1H2-Alletra-V001    tdvv v2    Yes   base ---      45 RW normal               99225    5242880
46  H1H2-Alletra-V003    tdvv v2    Yes   base ---      46 RW normal              642600    5242880
47  H2H1-Alletra-V004    tdvv v2    Yes   base ---      47 RW normal              215250    5242880
48  H1-Alletra-V100      tdvv v2    Yes   base ---      48 RW normal                6825    1048576
67  H1H2-Alletra-V005    tdvv v2    Yes   base ---      67 RW normal              772800    5242880
68  H1H2-Alletra-V007    tdvv v2    Yes   base ---      68 RW normal              326025    5242880
69  H1H2-Alletra-V009    tdvv v2    Yes   base ---      69 RW normal              266175    5242880
70  H1H2-Alletra-V011    tdvv v2    Yes   base ---      70 RW normal              281400    5242880
71  H1H2-Alletra-V013    tdvv v2    Yes   base ---      71 RW normal             1833300    5242880
72  H2H1-Alletra-V006    tdvv v2    Yes   base ---      72 RW normal              287175    5242880
73  H2H1-Alletra-V008    tdvv v2    Yes   base ---      73 RW normal              580125    5242880
74  H2H1-Alletra-V010    tdvv v2    Yes   base ---      74 RW normal              405825    5242880
75  H2H1-Alletra-V012    tdvv v2    Yes   base ---      75 RW normal             1249500    5242880
76  H2H1-Alletra-V014    tdvv v2    Yes   base ---      76 RW normal              181125    5242880
79  H1H2-scratchISO-V000 tdvv v2    Yes   base ---      79 RW normal                1050    1048576
 1 .srdata               full NA    NA    base ---       1 RW normal               92160      92160
 0 admin                 full NA    NA    base ---       0 RW normal               10240      10240
---------------------------------------------------------------------------------------------------
35 total                                                                         9186813  143233024
```



I am completely stumped by this. 
When i run the script parts manually in a python session on the checkmk server it works fine to parse the output.
Also as you notice the discovery using cmk command seems to produce some kind of outuput, but the checks go stale and become "vanished" when running a new discovery using the gui.


Crash Type	section
Operating System	Ubuntu 24.04.2 LTS
Checkmk Version	2.4.0p2
Python Version	3.12.9 (main, May 17 2025, 01:20:06) [GCC 13.2.0]

**Comment by Yogibaer75** [comment]
Today i had also a 3par system and fixed some of the parsing problems - i hope so :)
You can check with this version of the mkp. It was working on my system.
Commit 4690b303bd7d1ef51f8d4a5a9328c3161f9ffb99 should fix the parse problem.
Can you have a look at the power supplies? It was the only function i have not changed.
It is possible that i forgot to do this^^

**Comment by OlaHagbohm** [comment]
I finally got time to try out your updates, now i get no crash, but sadly no checks are being discovered.

```
Discovering services and host labels on: servername.domain.local
servername.domain.local:
+ FETCHING DATA
  Source: SourceInfo(hostname='servername.domain.local', ipaddress='10.1.2.3', ident='special_threepar_ssh', fetcher_type=<FetcherType.SPECIAL_AGENT: 6>, source_type=<SourceType.HOST: 1>)
[cpu_tracking] Start [7d89d4ce0320]
Read from cache: AgentFileCache(path_template=/omd/sites/monitor/tmp/check_mk/data_source_cache/special_threepar_ssh/servername.domain.local, max_age=MaxAge(checking=0, discovery=90.0, inventory=90.0), simulation=False, use_only_cache=False, file_cache_mode=1)
Calling: /omd/sites/monitor/local/lib/python3/cmk_addons/plugins/threepar_ssh/libexec/agent_threepar_ssh -u monitor -i showcage,showpd,showld,showvv,showps,shownode 10.1.2.3
Get data from program
[cpu_tracking] Stop [7d89d4ce0320 - Snapshot(process=posix.times_result(user=0.0, system=0.0, children_user=0.25, children_system=0.04, elapsed=6.839999999850988))]
+ PARSE FETCHER RESULTS
<<<3par_cage>>> / Transition NOOPParser -> HostSectionParser
<<<3par_pd>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_ld>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_vv>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_ps>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_node>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_cage>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_pd>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_ld>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_vv>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_ps>>> / Transition HostSectionParser -> HostSectionParser
<<<3par_node>>> / Transition HostSectionParser -> HostSectionParser
  HostKey(hostname='servername.domain.local', source_type=<SourceType.HOST: 1>)  -> Add sections: ['3par_cage', '3par_ld', '3par_node', '3par_pd', '3par_ps', '3par_vv']
Received no piggyback data
+ ANALYSE DISCOVERED HOST LABELS
Trying host label discovery with: 3par_cage, 3par_ld, 3par_node, 3par_pd, 3par_ps, 3par_vv
Trying host label discovery with:
SUCCESS - Found no new host labels
+ ANALYSE DISCOVERED SERVICES
+ EXECUTING DISCOVERY PLUGINS (6)
  Trying discovery with: 3par_pd, 3par_cage, 3par_node, 3par_ld, 3par_vv, 3par_ps
SUCCESS - Found no new services
```

**Comment by OlaHagbohm** [comment]
Realized the update you made did not match the output i am getting from the Alletra. Managed to update the files (except for 3par_ps, since that does not seem to exist on the alletra (shownode -ps does not exist))

```
+ ANALYSE DISCOVERED SERVICES
+ EXECUTING DISCOVERY PLUGINS (5)
  Trying discovery with: 3par_vv, 3par_pd, 3par_ps, 3par_ld, 3par_cage
  2 3par_cage
113 3par_ld
 32 3par_pd
 39 3par_vv
SUCCESS - Found 186 services
```

However, when looking at the host in checkmk i can not see the discovered services

<img width="729" height="135" alt="Image" src="https://github.com/user-attachments/assets/223ff380-32fc-4d59-b172-96b581eed0f4" />

and when attempting to run service discovery through the web gui i am met with this:

<img width="792" height="528" alt="Image" src="https://github.com/user-attachments/assets/f1168a47-fa6b-455c-a80e-f1a92d341b31" />
